### DISCUSSION TITLE
🚑 VERCEL DEPLOYMENT Error: The pattern "api/*.js" defined in `functions` doesn't match any Serverless Functions.

### DISCUSSION THREAD

**Discussion by HeavenlyEntity** [vercel/community#4391] [Help] [answered]
### Page to Investigate

Any Project deployment page

### Steps to Reproduce

# Foundation
The project s the latest version of nextjs `13.5.4`. Using the pages and `src` directory. Within a monorepo (turbo). I am a pro subscriber.

1. Create the `vercel.json` at the root of the project. and then add:

```json
{
  "framework": "nextjs",
  "devCommand": "next dev",
  "functions": {
    "api/*.js": {
      "memory": 3008,
      "maxDuration": 30
    }
  }
}
```
- I followed exactly how the docs explain it to work. https://vercel.com/docs/functions/serverless-functions/runtimes#maxduration

2. I already have a function defined in my project which is already defined. So now push an update.

![image](https://github.com/vercel/community/assets/39228141/2c794ed3-0578-42fd-9b8d-905272a20c1f)

3. Instant failure. Not sure what is happening. 
   - I wonder if it being in a monorepo hurts the finding of the files if its scanning at the root. When it don't need to be. It needs to look at the scope of the current directory.

Let me know what the cause of this is. I am going to have to remove this property from the `vercel.json` until this is fixed. Going to have to put a `config` export in each file.

**Comment by amyegan** [comment, marked as answer]
Hi @HeavenlyEntity. Since you are using Next.js, the function path may need to be adjusted. For example, I have an example project using the app router and I was able to set `maxDuration` for all of my Serverless Functions via vercel.json using this [`functions`](https://vercel.com/docs/projects/project-configuration#functions) config:
```
{
  "functions": {
    "app/api/**/route.ts": {
      "memory": 3008,
      "maxDuration": 30
    }
  }
}
```

This works for my test app because my function code is in `/app/api/hello/route.ts` and `/app/api/test/route.ts` starting from the root of the project code. You said you're using the `/src` directory, so you would need to add that to the beginning of the path.

<img width="245" alt="" src="https://github.com/vercel/community/assets/8247105/f6f64a4a-5b12-42e7-b950-be9ed205aa6b">

**Reply by HeavenlyEntity** [reply]
@amyegan this is perfect thank you for clarifying this! Where in the docs do they mention this? I would love to add it to my reading list and send it to my team 😄🙌

You're a lifesaver! 🚀

**Reply by amyegan** [reply]
Glad I could help! You can find more info about defining function key values in the `functions` documentation at https://vercel.com/docs/projects/project-configuration#functions

I'll take a look at what we can do to improve the [`maxDuration`](https://vercel.com/docs/functions/serverless-functions/runtimes#maxduration) documentation. Thanks for the feedback 😄

**Comment by dinakartumu** [comment]
@amyegan Do you know how to fix this issue or increase maxDuration for projects using redwood js framework.
using `{
  "functions": {
    "api/graphql": {
      "maxDuration": 120
    }
  }
}
` throws 'Error: The pattern "api/graphql" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.'

**Reply by amyegan** [reply]
You need to use the file path rather than the URL path. For example, `"api/function.js"` rather than `"api/function"`.

**Reply by dinakartumu** [reply]
i tried both "api/graphql.js" and "api/graphql.ts"  but i still get `Error: The pattern "api/graphql.js" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.`

The following represents the project structure recommended by Redwood JS.
![CleanShot 2023-10-12 at 03 42 47@2x](https://github.com/vercel/community/assets/3872554/06f521ff-5cbc-4f0c-9aab-9c53a1bba367)

**Reply by amyegan** [reply]
Based on that screenshot, it looks like the file path in your case is `"api/src/functions/graphql.ts"` or `"api/**/graphql.ts"`. I'm not familiar with your project so I may have gotten that wrong. 

The key is to use the real file path starting with the `/api` directory at the root of the project. You can find more examples at https://vercel.com/docs/projects/project-configuration#functions

**Reply by dinakartumu** [reply]
Still no luck. 
`Error: The pattern "api/src/functions/graphql.ts" defined in `functions` doesn't match any Serverless Functions inside the `api` directory`
`Error: The pattern "api/**/graphql.ts" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.`

below code is my vercel.json
```javascript
{
  "functions": {
    "api/src/functions/graphql.ts": {
      "maxDuration": 120
    }
  }
}

``` 

i also opened support ticket in vercel. https://vercel.com/teams/anilabsinc-s-team/support/cases/00164770.

**Reply by amyegan** [reply]
If you share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), I can take a closer look

**Reply by dinakartumu** [reply]
Here is the https://github.com/dinakaranilabsinc/redwood-maxduration-issue reproduciable example

**Reply by dinakartumu** [reply]
I have received a response from the Vercel support team. The issue has been resolved using the configuration provided below.
`{
  "functions": {
    "api/src/functions/graphql.*": {
      "maxDuration": 120,
      "runtime": "@vercel/redwood@2.0.5"
    }
  }
}`

**Reply by bencohen4** [reply]
@amyegan I am getting the same issue as @dinakartumu, on a RedwoodJS project, I am using the same `vercel.json` as in @dinakartumu's last response on Oct 31, 2023:

```
{ "functions": { "api/src/functions/graphql.*": { "maxDuration": 120, "runtime": "@vercel/redwood@2.0.5" } } }
```

It is not increasing the max duration at all - it is still 15 seconds. I'm on a Pro plan. Do you have any ideas for how I can go about debugging / fixing?

**Reply by swarnava** [reply]
Can you all try?

```json
{
  "functions": {
    "api/src/functions/*.*": {
      "maxDuration": 120,
      "runtime": "@vercel/redwood@2.0.5"
    }
  }
}
```

**Comment by tsamantanis** [comment]
@amyegan  Do you maybe know how to fix this issue or increase maxDuration for projects using Gatsbyjs? The file structure in the source is `src/api/example.js`

This is what my `vercel.json` looks like:
```
{
  "outputDirectory": "public",
  "functions": {
    "src/api/**.js": {
      "maxDuration": 30
    }
  }
}
```

I've tried a few different things but with no luck. Here are some screenshots from the source file structure and the output

**src**
![Screenshot 2023-11-21 at 23 25 57](https://github.com/vercel/community/assets/44101053/4fdd4dcd-09a3-47b8-ac6f-a50053c7b0c0)

**Output**
![Screenshot 2023-11-21 at 23 26 05](https://github.com/vercel/community/assets/44101053/6beed767-2232-4f7c-b1b7-6aeafe27d3dc)

**Reply by tsamantanis** [reply]
@amyegan Here is a minimal reproducible example of the issue https://github.com/tsamantanis/vercel-gatsby-serveless-functions-maxDuration-bug-repro

**Reply by amyegan** [reply]
Thanks for sharing that repo! I ran into some config errors deploying it. But I was able to get a quick example running with your `hello.js` function in the [Gatsby 5 Boilerplate](https://vercel.com/templates/gatsby/gatsbyjs-boilerplate) template. Since [Gatsby function routing](https://www.gatsbyjs.com/docs/reference/functions/routing/) supports both top-level and nested routes, I put my `/api` directory at the root of the repo.

You can see my example repo at https://github.com/amyegan/gatsbyjs-boilerplate-4391 (specifically [api/hello.js](https://github.com/amyegan/gatsbyjs-boilerplate-4391/blob/d9b44855a10e3d2d2135fafd072e4a4977213419/api/hello.js) and [vercel.json](https://github.com/amyegan/gatsbyjs-boilerplate-4391/blob/d9b44855a10e3d2d2135fafd072e4a4977213419/vercel.json))

<img width="615" alt="function details showing maximum execution duration set to 7 seconds" src="https://github.com/vercel/community/assets/8247105/e9207ef0-de1f-4b6f-aead-75f678a4f231">

**Reply by tsamantanis** [reply]
@amyegan  Hello thanks for the response! I moved the `api` folder to the root and, although it works for a simple route like the `hello.js` example, it is not possible to include packages like `axios` as the folder no longer has access to those. Any idea how to overcome this issue? Something to also note is that when using `gatsby develop` all functions throw a 404 since the `api` folder is no longer being read.

**Comment by imsamdez** [comment]
Sorry to revive this 3 months old discussion, but if you are using `pages` in nextjs, it seems that you need to include `pages/...` in your function path.

```JSON
{
  "functions": {
    "pages/api/*.js": {
      "memory": 3008,
      "maxDuration": 30
    }
  }
}
```

**Comment by smartraysam** [comment]
I am using remix framework
i have my serverless function in app/routes/api.chat.js "

am getting error
The pattern "app/routes/api.chat.tsx" defined in functions doesn't match any Serverless Functions inside the api directory.

please need help

**Reply by swarnava** [reply]
Hey, have you tried setting up directly within function? https://vercel.com/docs/functions/configuring-functions/duration#node.js-next.js-%3E=-13.5-or-higher-sveltekit-astro-nuxt-and-remix

**Reply by smartraysam** [reply]
Yes  I try add
export const config = {
  maxDuration: 5,
};

To the api route of the remix.
But fail to deploy. 
But deploy when I move it to the root.tsx.

But not sure if this will increase the duration for api routes
